README.md
# ktnsrm

Spectral power density models and Spectral Representation Method

## Introduction


## Usage

- Define a base Kanai Tajimi model;
- Define both separable and non-separable evlutionary power spectral density models;
- Generate sample realizations via the Spectral Representation Method;


## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`ktnsrm` was created by Y. Chen. It is licensed under the terms of the MIT license.



CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (20/02/2023)

- First release of `ktnsrm`!

tests/test_ktnsrm.py
from ktnsrm import ktnsrm


src/ktnsrm/utils.py
import matplotlib.pyplot as plt 
import numpy as np
import itertools
import pandas as pd
from matplotlib import mlab
from scipy.interpolate import griddata

def EPSD_show(Pxx, freqs, t_bins, format, title_name='the estimated spectra'):
        """Given the 3 elements returned by plt.specgram
        ie, (Pxx, freqs, t_bins)
        ----
        Change the colorbar
        """
        plt.figure(figsize=(6,4))
        if format=='2d':
            plt.pcolormesh(t_bins, freqs, Pxx, 
                    vmin=0, 
                    vmax=np.max(Pxx), 
                    shading='nearest', 
                    cmap='GnBu')
            plt.colorbar()
            plt.ylim([0, 15])
            plt.xlabel('time (s)')
            plt.ylabel('freq (HZ)')
            # plt.grid()
            plt.title(f'{title_name}')
        elif format=='3d':
            fig = plt.figure(figsize=(8,8))    
            ax = plt.axes(projection='3d')
            X, Y = np.meshgrid(t_bins, freqs)
            Z = Pxx
            ax.plot_surface(X, Y, Z, cmap='viridis')
            ax.set_xlabel('time (s)')
            ax.set_ylabel('frequency (hz)')
            ax.set_zlabel('PSD')
            ax.set_title(f'{title_name}')


def specgram3d(y, fs=200, title=None):
        """
        This func is borrowed elsewhere
        """
        ax = plt.axes(projection='3d')
        ax.set_title(title, loc='center', wrap=True)
        spec, freqs, t = mlab.specgram(y, Fs=fs, NFFT=256, noverlap=128)
        X, Y, Z = t[None, :], freqs[:, None], spec
        ax.plot_surface(X, Y, Z, cmap='viridis')
        ax.set_xlabel('time (s)')
        ax.set_ylabel('frequencies (Hz)')
        ax.set_zlabel('PSD')
#         ax.set_zlim([0, 0.015])
#         ax.set_ylim([0, 10])
#         ax.set_xlim([0, 14])
#         ax.view_init(20, 220)
        ax.invert_xaxis()
#         ax.invert_yaxis()
        return X, Y, Z


def simple_interpol2d(Pxx, freqs, t_bins):
    """Since the shape of estimated Swt spectra is smaller than 
    the one used to do SRM simulation, we interpolate it.
    TODO:
    -----
    The Swt should be interpolated to a given shape by (N1, duration * Fs)
    """
    # Use the same variables names as the Numpy example
    # Points, values, and then new coordinates
    points = list(itertools.product(t_bins, freqs))
    
    sample_df = pd.DataFrame()
    sample_df['X'] = [xy[0] for xy in points]
    sample_df['Y'] = [xy[1] for xy in points]
    # sample_df['value'] = values

    up_reversed = np.flipud(Pxx)
    values = np.ravel(up_reversed, order='F')

    x_min, x_max = 0.0, sample_df['X'].max()
    y_min, y_max = 0.0, sample_df['Y'].max()

    new_x_coord = np.linspace(x_min, x_max, 2500)
    new_y_coord = np.linspace(y_min, y_max, 1024)
    xx, yy = np.meshgrid(new_x_coord, new_y_coord)
    grid_z0 = griddata(points, values, xi=(xx, yy), method='nearest')
    print("double check the shape after interpolation:", grid_z0.shape)
    return grid_z0


src/ktnsrm/SpecRep.py
import numpy as np
import matplotlib.pyplot as plt

class SpecRep():

    def __init__(self, wg, zzeta, S0):
        """ initial the three parameter that define a KT model """ 

        self.wg = wg
        self.zzeta = zzeta
        self.S0 = S0


    def getKTspectrum(self, w):
        self._psd = self.S0 * (self.wg**4 + 4*(self.zzeta**2)*(self.wg**2)*(w**2)) / (((self.wg**2-w**2)**2) + 4*(self.zzeta**2)*(self.wg**2)*(w**2)) 


    @property
    def getPSD(self,):
        return self._psd



    #### time modulating function #####
    @staticmethod
    def envelop_tfunc1(t, b=4, c=0.8):
        """ set up a modulating function (from Liam 2015) """

        return b * (np.exp(-c * t) - np.exp(-2 * c * t))



    def sepEpsd(self, envelopfuncObj, t_axis):
        """ create a separable evolutionary spectum 
        
        Hint
        ----
        modulating_obj @ stationary_PSD


        Parameters
        ----------
        g(t)^2:
            the time modulating function;
        S(w):
            the stationary spectrm;
        """

        gt = envelopfuncObj(t_axis)
        gt2 = gt**2

        self._sepEPSD = np.outer(self._psd, gt2)
        print(r"the shape of the nonstationary spectra $S_{wt}$:", self._sepEPSD.shape)



    @staticmethod
    def nonsepEpsd(w_axis, t_axis):
        """ create a non-separable evollutionary spectrum 
        
        Parameters
        ----------        
        
        Hint
        ----
        Equation given in Liam 2015
        """


        a_list = []

        for w in w_axis:
            middleman = (w ** 2 / (5 * np.pi)) * np.exp(-0.15 * t_axis) * (t_axis ** 2) * np.exp( - ((w/(5 *np.pi)) ** 2) * t_axis)
            a_list.append(middleman)

        a_list = np.vstack(a_list)
        return a_list


    ##### displaying EPSD 2D/3D #####
    @staticmethod
    def EPSD_display(Pxx, freqs, t_bins, format, title_name='the estimated spectra'):
        """Given the 3 elements returned by plt.specgram
        
        Note
        ----
        To Compensate as one-sided plot


        ie, (Pxx, freqs, t_bins)
        ----
        Change the colorbar
        """

        if format=='2d':
            # hint: pcolormesh(t, f, EPSD)
            plt.pcolormesh(t_bins, freqs, Pxx, 
                    vmin=0, 
                    vmax=np.max(Pxx), 
                    shading='nearest', 
                    cmap='BuPu')
            plt.colorbar()
            plt.ylim([0, 100])
            plt.xlabel('Time (s)')
            plt.ylabel('Freq (rad)')
            plt.title(f'{title_name}')
        elif format=='3d':
            fig = plt.figure(figsize=(8, 6))    
            ax = plt.axes(projection='3d')
            X, Y = np.meshgrid(t_bins, freqs)
            Z = Pxx
            ax.plot_surface(X, Y, Z, cmap='BuGn')
            ax.set_xlabel('Time (s)')
            ax.set_ylabel('Frequency (rad)')
            ax.set_zlabel('PSD')
            ax.set_title(f'{title_name}')







src/ktnsrm/SpecRepMethod.py
"""
Introduction
============
* implementation for the Kanai Tajimi model and also
* implementation for Spectral Representation method by Shinozuka
"""

import numpy as np
import math
import matplotlib.pyplot as plt
import itertools
import pandas as pd
from scipy import interpolate
from scipy.interpolate import griddata
from utils import EPSD_show


class SRM:
    def __init__(self, wu=100, N1=1024, fs=200, duration=16):
        self.wu = wu
        self.N1 = N1        
        self.fs = fs
        self.duration = duration
    # print('the cutoff frequency wu set as:', wu)
        self.t_axis_4simu = np.arange(0, self.duration, 1 / self.fs)  # dt = 1 / Fs
        self.w_axis_4simu = np.arange(0, self.wu, self.wu/self.N1)

    # @property
    # def t_axis_4simu(self):
    #     """Create the t-axis of the sample simulation"""
    #     t_axis_simu = np.arange(0, self.duration, 1 / self.fs)  # dt = 1 / Fs
    #     return t_axis_simu


    # # set up the w (rad/s) axis
    # @property
    # def w_axis_4simu(self):
    #     w_axis = np.arange(0, self.wu, self.wu/self.N1)
    #     return w_axis 

    def _SpecRepsentation0(self, Sww, plot='y'):
        '''
        For now, this func received a spectra as argument,
        which may be obtained from 'getSww_from_a_model' func
        '''
    
        # create the t axis
        n = np.arange(self.N1)
        delta_w = self.wu / self.N1
        w_n = n * delta_w
        A_n = np.sqrt(2 * Sww * delta_w)
        phi_n = np.random.uniform(0, 2 * np.pi, self.N1)
    
        # Note that A0=0 or S(w0)=0
        sum = 0
        for i in range(1, self.N1):
            sum = sum + A_n[i] * np.cos(w_n[i] * self.t_axis_4simu + phi_n[i])
        simulation = sum * np.sqrt(2)
        # print('sampling frequency:', Fs)
        t_upper_limit = 2 * np.pi / (2 * self.wu)
        print("the lower limit of sampling frequency:", math.ceil(1 / t_upper_limit))
        print("the length of the simulation", simulation.shape)
        if plot == 'y':
            plt.plot(self.t_axis_4simu, simulation)
            plt.xlabel('time [s]')
            plt.ylabel('amp')
        return simulation


    @staticmethod
    def getSww_from_a_model(model, w_axis):
        return model(w_axis)



    def _interpo_spectra(self, Pxx, freqs, t_bins, plotting=True, format='2d', title_name='interpolated_spectra'):
    
        """
        Given Swt estimated by STFT, 
        ie. Pxx, freqs, t_bins, im = ax2.specgram(...)
        
        Follow the example procedures;
        # Points, values, and then new coordinates
        """
        points  = list(itertools.product(t_bins, freqs))
        sample_df = pd.DataFrame()
        sample_df['X'] = [xy[0] for xy in points]
        sample_df['Y'] = [xy[1] for xy in points]
    
        up_reversed = np.flipud(Pxx)
        values = np.ravel(up_reversed, order='F')
    
        x_min, x_max = 0.0, sample_df['X'].max()
        y_min, y_max = 0.0, sample_df['Y'].max()
    
        new_x_coord = np.linspace(x_min, x_max, self.duration * self.fs)
        new_y_coord = np.linspace(y_min, y_max, self.N1)
        xx, yy = np.meshgrid(new_x_coord, new_y_coord)
    
        grid_z0 = griddata(points, values, xi=(xx, yy), method='nearest')
        grid_z1 = np.flipud(grid_z0)
        
        self._interpolated_bundle = (grid_z1, new_y_coord, new_x_coord)
        # if plotting:
        #     # plotting the Swt on new axes
        #     EPSD_show(grid_z1, new_y_coord, new_x_coord, format=format, title_name=title_name)
        return grid_z1
    


    def _get_interpolated_spectra(self, format):
        EPSD_show(*self._interpolated_bundle, format=format, title_name='interpolated spectra')
        # plt.show()       

    

    def nonsta_simulation(self, Pxx, freqs, t_bins, plotting=False):
        """
        For given estimated spectra, do interpolation first and then do simulation;
        
        Paramters:
        ---------
        Pxx, freqs, t_bins: 
        Given the estimated spectra by STFT

        **kwargs :
        Used to control if showing the estimated spectra in 2d or 3d
        """
        interpolated_Swt = self._interpo_spectra(Pxx, freqs, t_bins)
        x = self._SpecRepsentation0(interpolated_Swt, plot='y')
        return x


    def SpecRepsentation3(self, Sww, t_bins, plot='y'):
        '''
        For now, this func received a spectra as argument,
        which may be obtained from 'getSww_from_a_model' func
        '''
    
        # create the t axis
        n = np.arange(self.N1)
        delta_w = self.wu / self.N1
        w_n = n * delta_w
        A_n = np.sqrt(2 * Sww * delta_w)
        phi_n = np.random.uniform(0, 2 * np.pi, self.N1)
    
        # Note that A0=0 or S(w0)=0
        sum = 0
        for i in range(1, len(Sww)):
            f = interpolate.interp1d(t_bins, 
                 A_n[i], 
                 kind='next', 
                 bounds_error=False, 
                 fill_value='extrapolate')

            A_new = f(self.t_axis_4simu)
            sum = sum + A_new * np.cos(w_n[i] * self.t_axis_4simu + phi_n[i])

        simulation = sum * np.sqrt(2)
        # print('sampling frequency:', Fs)
        t_upper_limit = 2 * np.pi / (2 * self.wu)
        print("the lower limit of sampling frequency:", math.ceil(1 / t_upper_limit))
        print("the length of the simulation", simulation.shape)
        if plot == 'y':
            plt.plot(self.t_axis_4simu, simulation)
        return simulation


    def SpecRepsentation2(self, Sww, t_bins, freqs, plot='y'):
        """
        Note
        ----

        From an estimated evolutionary spectra, do simulation.
        a estimated EPSD with shape, e.g. (65, 41);

        Based on 'wu', 'N1' -> the freq axis
        Based on 'Fs', 'duration' -> the time axis
        ergo, we will interpolate a spectra with shape (N1, duration * Fs)

        Parameters
        -----------
        """

        # create the t axis
        n = np.arange(self.N1)
        delta_w = self.wu / self.N1
        w_n = n * delta_w
        A_n = np.sqrt(2 * Sww * delta_w)
        phi_n = np.random.uniform(0, 2 * np.pi, self.N1)

        # # Note that A0=0 or S(w0)=0
        # sum = 0
        # if Sww.shape[-1] == self.t_axis_4simu.shape[0]:
        #     for i in range(1, self.N1):
        #         sum = sum + A_n[i] * np.cos(w_n[i] * self.t_axis_4simu + phi_n[i])
        # elif Sww.shape[-1] < self.t_axis_4simu.shape[0]:
            # assert(t_bins != None)

        # Hint: we will construct a 2D grid

        ''' before: x -> t_axis
                    y -> w_axis
            eg. (65, 41)
        
            After: (N1, duration * Fs) -> (1024, 2500)
        '''


        # xx, yy = np.meshgrid(t_bins, freqs)
        # z = Sww
        # f = interpolate.interp2d(t_bins, freqs, z, 
        #                          kind='next', 
        #                          bounds_error=False, 
        #                          fill_value='extrapolate')

        # xnew = self.t_axis_4simu
        # ynew = self.w_axis_4simu
        # znew = f(xnew, ynew)
        
        points = np.meshgrid(t_bins, freqs)
        values = Sww
        grid_x, grid_y = np.mgrid[self.t_axis_4simu, self.w_axis_4simu]
        grid_z0 = griddata(points, values, (grid_x, grid_y), method='nearest')

        return self.SpecRepsentation(grid_z0)





